File: app/src/main/java/com/example/tutorial/db/NotesProvider.java
package com.example.tutorial.db;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class NotesProvider extends ContentProvider {

    private static final UriMatcher URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);

    private static final int NOTES = 1;
    private static final int NOTE = 2;

    static {
        URI_MATCHER.addURI(NotesContract.AUTHORITY, "notes", NOTES);
        URI_MATCHER.addURI(NotesContract.AUTHORITY, "notes/#", NOTE);
    }

    private NotesDbHelper notesDbHelper;

    @Override
    public boolean onCreate() {
        notesDbHelper = new NotesDbHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        SQLiteDatabase db = notesDbHelper.getReadableDatabase();

        switch (URI_MATCHER.match(uri)) {
            case NOTES:
                if (TextUtils.isEmpty(sortOrder)) {
                    sortOrder = NotesContract.Notes.COLUMN_UPDATED_TS + " DESC";
                }

                return db.query(NotesContract.Notes.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder);

            case NOTE:
                String id = uri.getLastPathSegment();

                if (TextUtils.isEmpty(selection)) {
                    selection = NotesContract.Notes._ID + " = ?";
                    selectionArgs = new String[]{id};
                } else {
                    selection = selection + " AND " + NotesContract.Notes._ID + " = ?";

                    String[] newSelectionArgs = new String[selectionArgs.length + 1];

                    System.arraycopy(selectionArgs, 0, newSelectionArgs, 0, selectionArgs.length);

                    newSelectionArgs[newSelectionArgs.length - 1] = id;

                    selectionArgs = newSelectionArgs;
                }

            default:
                return null;
        }
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        switch (URI_MATCHER.match(uri)) {
            case NOTES:
                return NotesContract.Notes.URI_TYPE_NOTE_DIR;
            case NOTE:
                return NotesContract.Notes.URI_TYPE_NOTE_ITEM;
            default:
                return null;
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        SQLiteDatabase db = notesDbHelper.getWritableDatabase();

        switch (URI_MATCHER.match(uri)) {
            case NOTES:
                long rowId = db.insert(NotesContract.Notes.TABLE_NAME,
                        null,
                        values);

            default:
                return null;
        }
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: app/src/main/java/com/cruelgeroi/apphub/db/NotesProvider.kt
package com.cruelgeroi.apphub.db

import android.content.ContentProvider
import android.content.UriMatcher
import android.text.TextUtils
import android.content.ContentValues
import android.content.ContentUris
import android.database.Cursor
import android.net.Uri
import android.provider.BaseColumns

class NotesProvider : ContentProvider() {
    companion object {
        private val URI_MATCHER = UriMatcher(UriMatcher.NO_MATCH)
        private const val NOTES = 1
        private const val NOTE = 2

        init {
            URI_MATCHER.addURI(NotesContract.AUTHORITY, "notes", NOTES)
            URI_MATCHER.addURI(NotesContract.AUTHORITY, "notes/#", NOTE)
        }
    }

    private var notesDbHelper: NotesDbHelper? = null
    override fun onCreate(): Boolean {
        notesDbHelper = NotesDbHelper(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        var selection = selection
        var selectionArgs: Array<String>? = selectionArgs
        var sortOrder = sortOrder
        val db = notesDbHelper!!.readableDatabase

        return when (URI_MATCHER.match(uri)) {
            NOTES -> {

                if (TextUtils.isEmpty(sortOrder)) {
                    sortOrder = NotesContract.Notes.COLUMN_UPDATED_TS + " DESC"
                }

                db.query(
                    NotesContract.Notes.TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs,
                    null,
                    null,
                    sortOrder
                )
            }
            NOTE -> {
                var id: String = uri.lastPathSegment.toString()
                if (TextUtils.isEmpty(selection)) {
                    selection = BaseColumns._ID + " = ?"
                    selectionArgs = arrayOf(id)
                } else {
                    selection = selection + " AND " + BaseColumns._ID + " = ?"
                    val newSelectionArgs = arrayOf(
                        (selectionArgs!!.size + 1).toString()
                    )
                    System.arraycopy(
                        selectionArgs,
                        0,
                        newSelectionArgs,
                        0,
                        selectionArgs.size
                    )
                    newSelectionArgs[newSelectionArgs.size - 1] = id
                    selectionArgs = newSelectionArgs
                }
                null
            }
            else -> null
        }
    }

    override fun getType(uri: Uri): String? {
        return when (URI_MATCHER.match(uri)) {
            NOTES -> NotesContract.Notes.URI_TYPE_NOTE_DIR
            NOTE -> NotesContract.Notes.URI_TYPE_NOTE_ITEM
            else -> null
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val db = notesDbHelper!!.writableDatabase
        return when (URI_MATCHER.match(uri)) {
            NOTES -> {
                val rowId = db.insert(
                    NotesContract.Notes.TABLE_NAME,
                    null,
                    values
                )
                if (rowId > 0) {
                    val noteUri = ContentUris.withAppendedId(NotesContract.Notes.URI, rowId)
                    context!!.contentResolver.notifyChange(uri, null)
                    return noteUri
                }
                null
            }
            else -> null
        }
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }
}